src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -0,0 +1,543 @@
+use sink::EventSink;
+pub use state::WinitState;
+
+pub struct EventLoopWindowTarget<T> {
+    /// Wayland display.
+    pub display: Display,
+
+    /// Environment to handle object creation, etc.
+    pub env: Environment<WinitEnv>,
+
+    /// Event loop handle.
+    pub event_loop_handle: calloop::LoopHandle<WinitState>,
+
+    /// Output manager.
+    pub output_manager: OutputManager,
+
+    /// Wayland source.
+    pub wayland_source: Rc<calloop::Source<WaylandSource>>,
+
+    /// Window map.
+    pub window_map: Rc<RefCell<HashMap<WindowId, WindowHandle>>>,
+
+    /// Pending window updates.
+    pub window_updates: Rc<RefCell<HashMap<WindowId, WindowUpdate>>>,
+
+    /// Window event sink.
+    pub event_sink: Rc<RefCell<EventSink>>,
+
+    /// A proxy to wake up event loop.
+    pub event_loop_awakener: calloop::ping::Ping,
+
+    /// The available windowing features.
+    pub windowing_features: WindowingFeatures,
+
+    /// Theme manager to manage cursors.
+    ///
+    /// It's being shared amoung all windows to avoid loading
+    /// multiple similar themes.
+    pub theme_manager: ThemeManager,
+
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> EventLoopWindowTarget<T> {
+    pub fn get_winit_state(&self) -> WinitState {
+        let event_sink = self.event_sink.clone();
+        let window_updates = self.window_updates.clone();
+        // let event_loop_awakener = self.event_loop_awakener.clone();
+        let window_map = self.window_map.clone();
+        WinitState::new(event_sink, window_updates, window_map)
+    }
+}
+
+pub struct EventLoop<T: 'static> {
+    /// Event loop.
+    event_loop: calloop::EventLoop<WinitState>,
+
+    /// Wayland display.
+    display: Display,
+
+    /// Pending user events.
+    pending_user_events: Rc<RefCell<Vec<T>>>,
+
+    /// Source of user events.
+    _user_events_source: calloop::Source<calloop::channel::Channel<T>>,
+
+    /// Sender of user events.
+    user_events_sender: calloop::channel::Sender<T>,
+
+    /// Wayland source of events.
+    wayland_source: Rc<calloop::Source<WaylandSource>>,
+
+    /// Source of event loop wake ups, where we handle pending window requests.
+    _event_loop_awakener_source: calloop::Source<calloop::ping::PingSource>,
+
+    /// Window target.
+    window_target: RootEventLoopWindowTarget<T>,
+
+    /// Output manager.
+    _seat_manager: SeatManager,
+}
+
+impl<T: 'static> EventLoop<T> {
+    pub fn new() -> Result<EventLoop<T>, Box<dyn Error>> {
+        // Connect to wayland server and setup event queue.
+        let display = Display::connect_to_env()?;
+        let mut event_queue = display.create_event_queue();
+        let display_proxy = display.attach(event_queue.token());
+
+        // Setup environment.
+        let env = Environment::init(&display_proxy, WinitEnv::new());
+
+        event_queue.sync_roundtrip(&mut (), |_, _, _| unreachable!())?;
+        event_queue.sync_roundtrip(&mut (), |_, _, _| unreachable!())?;
```

Comment:
Those things setup globals you asked for in environment, in general sctk does that behind the scenes with its macros, but I went for more manual control in winit for such things, since I don't need some parts of default environemnt, etc.

Reply:
How much of that do we actually not need and what difference does it make for winit? Since you're heavily involved with sctk you probably are aware of when this is needed, but I think it causes an extreme risk and maintenance burden should there ever be someone else maintaining it (and just reviews alone are more difficult of course).

Reply:
It includes clipboards protocols by default, and we can't really handle or exclude those in winit, moreover it'll break GNOME, since it really don't like binding clipboard manager more than once in application. Also, manual managing is the way to go most of the time if you're using a lot of magic protocols, since you setup it once, and you can't really fail here tbqh. I'll add a comment why those roundtrips are there, but for someone who is working with sctk a lot it's kind of expected.

Reply:
>and just reviews alone are more difficult of course

You may look at a commit that added IME and so one more field there, it's really easy to review, since you can't access IME without adding object into environment, for some that is completely unfamiliar how Wayland works it could be not really obvious though, but if you work with it you know that you should bind global, and that's pretty explicitly with sctk's environment.

Reply:
> It includes clipboards protocols by default, and we can't really handle or exclude those in winit

That seems like an SCTK issue that should be brought up to SCTK.

> Also, manual managing is the way to go most of the time if you're using a lot of magic protocols

Why are we using a lot of magic protocols? We shouldn't be.

> I'll add a comment why those roundtrips are there, but for someone who is working with sctk a lot it's kind of expected.

I disagree with the notion that someone has to have a lot of experience with SCTK to work on winit's Wayland system. That is something that I think is ***extremely*** dangerous to the maintainability of the platform and would strike me as so much of a concern, that I'm not sure I'm comfortable with approving this approach.

> You may look at a commit that added IME and so one more field there, it's really easy to review, since you can't access IME without adding object into environment, for some that is completely unfamiliar how Wayland works it could be not really obvious though, but if you work with it you know that you should bind global, and that's pretty explicitly with sctk's environment.

No, this is not easy to review or obvious at all. Having to call `sync_roundtribe`, **especially multiple times in a row** is completely non-obvious to anyone reading the code. That kind of API is just completely unacceptable from a maintainability POV. The code is not self-descriptive at all and there will be problems if things even just get reordered a little bit. You literally couldn't build anything more fragile.

Reply:
>That seems like an SCTK issue that should be brought up to SCTK.

I guess. We've discussed the `default_env!` macro, but we need to extend it anyway. Like the macro itself isn't ideal, but it's know problem, there're no issue about that.

Reply:
>Why are we using a lot of magic protocols? We shouldn't be.

All protocols that are their are required and I've documented why certain things is required for winit, right? If you need more docs for what we're using let me know, but it's a `minimum` that we need in winit to get things working.

>No, this is not easy to review or obvious at all. Having to call sync_roundtribe, especially multiple times in a row is completely non-obvious to anyone reading the code. 

A common thing on Wayland even for C code though.

>That kind of API is just completely unacceptable from a maintainability POV. The code is not self-descriptive at all and there will be problems if things even just get reordered a little bit. You literally couldn't build anything more fragile if you wanted to.

SCTK can't call `sync_roundtrip` for you, it generate those in `init_default_env` macro, so SCTK macros should be improved to handle those.

Reply:
> A common thing on Wayland even for C code though.

But this isn't a C codebase?

> SCTK can't call sync_roundtrip for you, it generate those in init_default_env macro, so SCTK macros should be improved to handle those.

That sounds like SCTK's problem to figure out? Because right now there is zero indication from SCTK whether sync_roundtrip needs to be called or not. Especially calling it multiple times is just completely arbitrary and makes no sense.

Reply:
>That sounds like SCTK's problem to figure out? Because right now there is zero indication from SCTK whether sync_roundtrip needs to be called or not. Especially calling it multiple times is just completely arbitrary and makes no sense.

It's right in the docs. SCTK docs are quite verbose on what you should do tbqh.

https://docs.rs/smithay-client-toolkit/0.11.0/smithay_client_toolkit/environment/struct.Environment.html#method.init

Reply:
Ignoring the typos, some random sidenote in a method is hardly a good API. I'm not saying it's the fault of your winit implementation, but this API (SCTK) is a huge maintenance risk if you ask me.

Having to read up on docs on what you're allowed/not allowed/supposed to do is some serious C level of API with so little help from the typesystem. To me that's unacceptable in Rust.

Reply:
>Ignoring the typos, some random sidenote in a method is hardly a good API. I'm not saying it's the fault of your winit implementation, but this API (SCTK) is a huge maintenance risk if you ask me.

I guess, it could be improved in sctk, like it's you know evolving and we're free to send patches there. Just saying that it takes time to update. It has a macro called `init_default_env!` that does every that roundtrip stuff for you, but it doesn't have more genericish way to setup env. I guess it could take event queue as an argument, so it'll dispatch for you, but that's up to sctk tbqh, and nothing I can do about in winit.

>Having to read up on docs on what you're allowed/not allowed/supposed to do is some serious C level of API with so little help from the typesystem. To me that's unacceptable in Rust.

Yeah, it feels C-ish in such cases, but if you don't read docs of a lib you're using how are you even writing the code and understand what to use?

Reply:
> nothing I can do about in winit

Making the commitment to add support for an API that is not maintainable is something you have to do in winit though. If the Wayland bindings are not usable, I'm not sure it's a good idea to bind to them and pull all that maintenance burden into Alacritty.

This might still be an improvement compared to the rest of winit, but that's solely due to its already less than ideal quality. I personally wouldn't be comfortable adding this myself with the code quality that comes with it.

> Yeah, it feels C-ish in such cases, but if you don't read docs of a lib you're using how are you even writing the code and understand what to use?

The typesystem alone should be sufficient for me to make things work. I generally never read through the entire documentation text but piece together code examples. If your API does not enforce doing it right, it's a useless API that just wastes my time having to read through documentation for every single item to be able to use it. This kind of stuff is exactly the reason why a lot of C code is so unmaintainable.

Reply:
>Making the commitment to add support for an API that is not maintainable is something you have to do in winit though. If the Wayland bindings are not usable, I'm not sure it's a good idea to bind to them and pull all that maintenance burden into Alacritty.

>This might still be an improvement compared to the rest of winit, but that's solely due to its already less than ideal quality. I personally wouldn't be comfortable adding this myself with the code quality that comes with it.

According to you if I'm using plain C API in winit like we do with X11 it shouldn't be in winit in a first place? It's just a mess if you look at X11 code, and the same applies to other platforms, here we have one minor thing like calling 2 sync roundtrips and it 'completely unmaintainable'. Only one piece of code requires such magic and I'm not sure what kind of type system over basically raw Wayland mechanism you want here. You can't really provide event queue inside of the `init` tbqh, since you may want to have different callbacks for certain things. I'm fine with maintaining that code, since it's very clear for me how it works and it's very easy to add literally anything in the current Wayland backend, unlike other backends.

>The typesystem alone should be sufficient for me to make things work. I generally never read through the entire documentation text but piece together code examples. If your API does not enforce doing it right, it's a useless API that just wastes my time having to read through documentation for every single item to be able to use it. This kind of stuff is exactly the reason why a lot of C code is so unmaintainable.

Ofc and then you have async protocol on you with pretty strict requirements on what to do, how to do, and when to do. I can surely work in SCTK to make it more automatic for the end users, so the won't need to carry about stuff like initializing their Environment in a manual way, since common sctk users are using `init_default_environment!` call which already have setup everything for you including those 2 sync roundtrips. And that is what every example is doing https://github.com/Smithay/client-toolkit/blob/aac3c503242c8a2a9f37f4a2231e7b540e3a575c/examples/kbd_input.rs#L21 .

Reply:
> According to you if I'm using plain C API in winit like we do with X11 it shouldn't be in winit in a first place?

I've worked with X11 myself and I have not run into issues like these. The winit implementation being less than maintainable shouldn't be a surprise though, I am not particularly fond of its code quality.

> what kind of type system over basically raw Wayland mechanism you want here

I still don't see why we need to use "basically raw Wayland" in winit. My goal is exactly to not just have raw Wayland since it is completely useless to winit and unnecessarily difficult to maintain.

> I'm fine with maintaining that code, since it's very clear for me how it works and it's very easy to add literally anything in the current Wayland backend, unlike other backends.

vberger has already stepped away from winit. Having only a single person capable of maintaining a backend is not sufficient to ensure longevity. This is not about **you** being able to maintain it, but at least one other person capable of doing so. I currently do not see any other person being capable of even reviewing these PRs if vberger doesn't help out who is just capable of reviewing it since he wrote the API himself.

> I can surely work in SCTK to make it more automatic for the end users, so the won't need to carry about stuff like initializing their Environment in a manual way, since common sctk users are using init_default_environment! call which already have setup everything for you including those 2 sync roundtrips.

So let's do that then? Clearly if it can be done with the default environment, it should be possible. It just seems to me like the entire environment API is lacking in SCTK.

Reply:
>I still don't see why we need to use "basically raw Wayland" in winit. My goal is exactly to not just have raw Wayland since it is completely useless to winit and unnecessarily difficult to maintain.

The only raw Wayland part that is hard to maintain is that `new` method in `EventLoop`, since you're required to read the docs on how to bootstrap everything. Out side of it you don't need anything.

>vberger has already stepped away from winit. Having only a single person capable of maintaining a backend is not sufficient to ensure longevity. This is not about you being able to maintain it, but at least one other person capable of doing so. I currently do not see any other person being capable of even reviewing these PRs if vberger doesn't help out who is just capable of reviewing it since he wrote the API himself.

The only thing that requires review from someone that read the docs in SCTK is the creation of event loop and creation of a window in winit, since that sync roundtrip isn't something that you usually have in normal clients, since you don't draw before configure anyway. I can maybe work on removing it, but I can't come up with something that works well.

>So let's do that then? Clearly if it can be done with the default environment, it should be possible. It just seems to me like the entire environment API is lacking in SCTK.

Well, that's my plan to work on SCTK and winit, but I don't want to hold that PR any longer, since the old backend is impossible to maintain even for me, and adding anything to it is literally insanity.

I really don't feel like continuing the discussion here, since I'm not sure that something will be changed, and it's basically just wasting of our time.